packages/flutter/lib/src/services/text_formatter.dart

```diff
@@ -121,6 +121,94 @@ class _SimpleTextInputFormatter extends TextInputFormatter {
   }
 }
 
+// A mutable, half-open range [`base`, `extent`) within a string.
+class _MutableTextRange {
+  _MutableTextRange(this.base, this.extent);
+
+  static _MutableTextRange? fromComposingRange(TextRange range) {
+    return range.isValid && !range.isCollapsed
+      ? _MutableTextRange(range.start, range.end)
+      : null;
+  }
+
+  static _MutableTextRange? fromTextSelection(TextSelection selection) {
+    return selection.isValid
+      ? _MutableTextRange(selection.baseOffset, selection.extentOffset)
+      : null;
+  }
+
+  /// The start index of the range, inclusive.
+  ///
+  /// The value of [base] should always be greater than or equal to 0, and can
+  /// be larger than, smaller than, or equal to [extent].
```

Comment:
Do we need to worry about inverted selection?

Reply:
Oh right I think I just copied the documentation over. No I think the indices are processed separately and it doesn't care if base > extent or not. I'll add a test for this.

Reply:
Oh nvm the documentation was right, it works fine with inverted selections. Many test cases have the selection range inverted.